### PR TITLE
feat(feedback): add User Feedbacks as Events

### PR DESCRIPTION
**Pull request by JoshFerge**
Relay changes for the revamped User Feedback feature. User Feedbacks (Known as `UserFeedbackV2`), will be events that flow through our system instead of attachments. See [here](https://www.notion.so/sentry/User-Feedback-Beta-11d68db039bc436bae017ad56c0c20f9?pvs=4) for more information on the tech spec (and the link at the bottom for what we do once they're ingested).

Note that within this code, we're calling it `UserReportV2`. This is to avoid confusion with the existing `UserReport`. A PR to shim the previous `UserReports` as new `UserReportV2`'s will follow, at which point we'll rename this in the code to be `UserFeedback`. The string `feedback` is used for the context and ItemType / EventType parsing.

- [x] Adds a feature flag for ingest `UserReportV2Ingest`. This is backed by options https://github.com/getsentry/getsentry/issues/11573
- [x] Creates a new EventType `UserReportV2`
- [x] Creates a new ItemType `UserReportV2`
- [x] Creates a new DataCategory `UserReportV2`
- [x] Adds a new context `UserReportV2Context` that for now contains two properties, `contact_email` (no pii stripping), and the `message` which contains the feedback text.

### CODE REVIEW

File: relay-base-schema/src/data_category.rs
Comment (iker-barriocanal) [inline relay-base-schema/src/data_category.rs]: Please also update the `from_name` function above.

File: relay-base-schema/src/events.rs
Comment (iker-barriocanal) [inline relay-base-schema/src/events.rs]: As per Relay's coding guidelines, documentation should have one line followed by additional paragraphs:
```suggestion
    /// User feedback payload.
    ///
    /// TODO(Jferg): Change this to UserFeedback once old UserReport logic is deprecated.
    UserReportV2,
```

File: relay-base-schema/src/data_category.rs
Comment (iker-barriocanal) [inline relay-base-schema/src/data_category.rs:65]: Depending on when UserReport -> UserFeedback transition happens, we may need to support both alternatives to avoid future breaking changes. Have we considered this? Is there any ETA on the deprecation?

File: relay-dynamic-config/src/feature.rs
Comment (iker-barriocanal) [inline relay-dynamic-config/src/feature.rs]:
```suggestion
    /// Enables new User Feedback ingest.
    ///
    /// TODO(jferg): rename to UserFeedbackIngest once old UserReport logic is deprecated.
```

File: relay-event-schema/src/protocol/contexts/mod.rs
Comment (iker-barriocanal) [inline relay-event-schema/src/protocol/contexts/mod.rs:30]: `*` exports above aren't ideal, could we only export the types we need?

File: relay-server/src/actors/processor.rs
Comment (iker-barriocanal) [inline relay-server/src/actors/processor.rs:1729]: Returning an error here means the envelope will not be processed, even if it also contains metrics. Should we only call `event_from_json_payload` if the flag is enabled?

File: relay-server/src/envelope.rs
Comment (iker-barriocanal) [inline relay-server/src/envelope.rs:564]: This means no outcomes will be generated. Is this intended?

**Review by iker-barriocanal — CHANGES_REQUESTED** [review]
Leaving some questions/points from a preliminary review:
- Since we're adding a new data category, we should update the C ABI headers -- `make header` in `relay-cabi` should work. We should also update the py/Changelog.
- What's the plan for current feedback items that live as attachments?
- Should these new feedback items go into a new Kafka topic?
- See my comment below on UserReport -> UserFeedback transition.

File: relay-base-schema/src/data_category.rs
Comment (jjbayer) [reply relay-base-schema/src/data_category.rs:65]: Why not use `UserFeedback` right away? I agree with Iker that renaming later could bring extra complexity and possible compatibility issues.

File: relay-base-schema/src/events.rs
Comment (jjbayer) [inline relay-base-schema/src/events.rs:78]: Similar question: If the string representation is `"feedback"`, why not call it `UserFeedback` in code?

File: relay-server/src/actors/processor.rs
Comment (jjbayer) [reply relay-server/src/actors/processor.rs:1729]: @iker-barriocanal we split event-like items into a separate envelope, so that should be fine.

Raising an error here will log an `invalid` outcome, but only if `outcome_category()` returns `Some` (see discussion below).

File: relay-base-schema/src/data_category.rs
Comment (jjbayer) [inline relay-base-schema/src/data_category.rs:65]: Separate question: Why do we need a new data category, if user feedback items are converted to events (i.e. `Error` category)?

Comment (JoshFerge) [reply relay-base-schema/src/data_category.rs:65]: they are not going to be Error category events. They will be billed separately.

File: relay-base-schema/src/events.rs
Comment (JoshFerge) [reply relay-base-schema/src/events.rs:78]: when discussing with @jan-auer over initial call, he called out to not have it so there are "synonyms" in the code. I can change to UserFeedback if that is preferred.

File: relay-server/src/envelope.rs
Comment (JoshFerge) [reply relay-server/src/envelope.rs:564]: for now, yes. will change this once we have data flowing through system and we're closer to alpha etc.

File: relay-base-schema/src/data_category.rs
Comment (JoshFerge) [reply relay-base-schema/src/data_category.rs:65]: when discussing with @jan-auer over initial call, he called out to not have it so there are "synonyms" in the code. I can change to UserFeedback if that is preferred.

File: relay-event-schema/src/protocol/contexts/mod.rs
Comment (JoshFerge) [reply relay-event-schema/src/protocol/contexts/mod.rs:30]: should i include that refactor in this PR, or could I follow up with that? not sure how you all like to split changes like that into separate PRs or not.

**Comment by JoshFerge** [comment]
> * Since we're adding a new data category, we should update the C ABI headers -- `make header` in `relay-cabi` should work. We should also update the py/Changelog.

done.

> * What's the plan for current feedback items that live as attachments?

see PR description:
> A PR to shim the previous UserReports as new UserReportV2's will follow, at which point we'll rename this in the code to be UserFeedback.

> * Should these new feedback items go into a new Kafka topic?

No -- we'll process them in our ingest consumer conditioned on the event type. this was chosen over a separate Kafka topic in conversation w/ @jan-auer last week.

> * See my comment below on UserReport -> UserFeedback transition.

Replied, when Jan and I spoke, there was a decision to avoid synonyms in the code. can change if ya'll would like -- I don't have a strong opinion here. also see PR description:

> at which point we'll rename this in the code to be UserFeedback

File: relay-base-schema/src/data_category.rs
Comment (jjbayer) [reply relay-base-schema/src/data_category.rs:65]: I think more important than the internal naming is making sure that the serialized string representations in the public interface do not have to be changed, e.g `"user_report_v2"` for the data category.

The internal names of the types are less important, but I did a case insensitive search for "feedback" and found only a few references in comments, so I would personally go for the new naming straight away, but I'll leave it up to you. cc @jan-auer

File: CHANGELOG.md
Comment (iker-barriocanal) [inline CHANGELOG.md]: Please also update py/changelog

**Review by iker-barriocanal — COMMENTED** [review]
Pending the `relay-cabi` update, LGTM. Thanks a lot for working on this PR!

File: CHANGELOG.md
Comment (jjbayer) [reply CHANGELOG.md]: Also, please move this to the `## Unreleased` section at the top of this file.

File: py/CHANGELOG.md
Comment (jjbayer) [inline py/CHANGELOG.md]: Did the autoformatter run wild here?

File: relay-event-schema/src/protocol/contexts/user_report_v2.rs
Comment (jjbayer) [inline relay-event-schema/src/protocol/contexts/user_report_v2.rs]: I think in the serialized format we should use the future-proof name, i.e. `"feedback"` or `"userfeedback"`. You can accomplish that by adding a `#[metastructure(tag = "feedback")]` attribute to the variant.

File: py/CHANGELOG.md
Comment (JoshFerge) [reply py/CHANGELOG.md]: fixed! not sure why my markdown formatter does this.

File: relay-event-schema/src/protocol/contexts/user_report_v2.rs
Comment (JoshFerge) [reply relay-event-schema/src/protocol/contexts/user_report_v2.rs]: good call. just made that change.

**Review by jjbayer — APPROVED** [review]

**Comment by JoshFerge** [comment]
@iker-barriocanal could you take a look again? need your approval since you requested changes.

File: relay-base-schema/src/data_category.rs
Comment (iker-barriocanal) [inline relay-base-schema/src/data_category.rs:118]: Please add this string to the `from_name` above before merging, otherwise Relay won't recognize the data category when deserializing.

**Review by iker-barriocanal — APPROVED** [review]

**Pull request merged**
